[dtensor] improve new factory strategy

tianyu-l opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122997
* #122996
* __->__ #122995

Previously, the new tensor out of the "new factory" all become replicated.
With this PR, if the new tensor has the same shape as the old tensor **and** the shape can be evenly sharded, then the old spec is inherited and preferred.

To accommodate this when the old tensor has sharded placements, the input args for local computation (size, stride) need to be adjusted.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225 @chauhang @d4l3k @rohan-varma

tianyu-l committed:
[dtensor] improve new factory strategy

[ghstack-poisoned]

Skylion007 commented on torch/distributed/_tensor/ops/tensor_ops.py:
```diff
@@ -181,15 +181,34 @@ def new_factory_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
         )
 
+        # if input shape is the same as output shape, we can just propagate input sharding
+        if list(input_shape) == list(output_shape):
```
NIT: tuples are cheaper to construct and would satisfy the equality here.

Skylion007 commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -190,6 +190,27 @@ def default_tensor(spec: DTensorSpec) -> torch.Tensor:
             else:
+                # local_tensor_args need to be modified for new factory ops, potentially
+                # TODO: adjust the stride (args[2]) for aten.new_empty_strided.default
+                if op_call in [
```
Replace list with tuple here.

Skylion007 reviewed (commented):
Nits

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -190,6 +190,27 @@ def default_tensor(spec: DTensorSpec) -> torch.Tensor:
                 ):
                     local_results = op_call(*local_tensor_args, **op_info.local_kwargs)
             else:
+                # local_tensor_args need to be modified for new factory ops, potentially
```
I think we should not do this in dispatch.py, instead we should do this in the sharding propagation, the reason is that adjusting to the local shape looks like a resharding (that's actually what we already did too) operation where we should simply generate a redistribute_schema that involves the input args that are adjusted already to local_shape. what we need to do:

1. maintain a dict of ops that map to which arg need to be adjusted to local (shape/stride) in the Propagator
2. after we get the output OpStrategy, we examine if the op is in the dict of ops, if it is, we adjust the corresponding args to local shape/stride when generating a redistribute_schema
3. return this as a redistribute_schema then dispatch.py would handle this automatically

wanchaol reviewed (changes requested):
Thanks for the improvements! I have some comments about where to do this cleanly

XilunWu commented on test/distributed/_tensor/test_tensor_ops.py:
```diff
@@ -321,21 +321,29 @@ def _test_op(self, mesh, op_call, *args, **kwargs):
+
+            with comm_mode:
+                new_full_same_dt = input_dt.new_full((12, 8), 42.0)
+                # new_full_same_dt creates a tensor with the same placement as input_dt,
```
what is the difference between `new_full_diff_dt` and `new_full_same_dt` test?

tianyu-l replied:
The difference is in shapes. The former has output tensor with a different shape than the input tensor; the latter's output tensor has the same shape as input.

tianyu-l committed:
Update on "[dtensor] improve new factory strategy"


Previously, the new tensor out of the "new factory" all become replicated.
With this PR, if the new tensor has the same shape as the old tensor, then the old spec is inherited and preferred.

To accommodate this when the old tensor has sharded placements, the input args for local computation (most prominently the shape) need to be adjusted.

wanchaol to discuss: for `new_empty_strided`, the local arg `stride` may also need to be adjusted. I found this unclear when the shape cannot be evenly sharded.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] improve new factory strategy"


Previously, the new tensor out of the "new factory" all become replicated.
With this PR, if the new tensor has the same shape as the old tensor **and** the shape can be evenly sharded, then the old spec is inherited and preferred.

To accommodate this when the old tensor has sharded placements, the input args for local computation (size, stride) need to be adjusted.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] improve new factory strategy"


Previously, the new tensor out of the "new factory" all become replicated.
With this PR, if the new tensor has the same shape as the old tensor **and** the shape can be evenly sharded, then the old spec is inherited and preferred.

To accommodate this when the old tensor has sharded placements, the input args for local computation (size, stride) need to be adjusted.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/sharding_prop.py:
```diff
@@ -417,3 +436,42 @@ def _select_strategy(self, strategy: OpStrategy) -> PlacementStrategy:
+            assert isinstance(stride, list)
+
+            stride_divisors = [1] * len(stride)
+            for i, mesh_idx in enumerate(spec.dim_map):
```
can we make this a util function? I think this would be useful for some other features in the future, similar to this https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/_utils.py#L17

Also please not use `dim_map`, at least that's a field we would need to do some refactor

tianyu-l committed:
Update on "[dtensor] improve new factory strategy"


Previously, the new tensor out of the "new factory" all become replicated.
With this PR, if the new tensor has the same shape as the old tensor **and** the shape can be evenly sharded, then the old spec is inherited and preferred.

To accommodate this when the old tensor has sharded placements, the input args for local computation (size, stride) need to be adjusted.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm! some nits inlined

tianyu-l committed:
Update on "[dtensor] improve new factory strategy"


Previously, the new tensor out of the "new factory" all become replicated.
With this PR, if the new tensor has the same shape as the old tensor **and** the shape can be evenly sharded, then the old spec is inherited and preferred.

To accommodate this when the old tensor has sharded placements, the input args for local computation (size, stride) need to be adjusted.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

tianyu-l commented:
@pytorchbot merge
